File: c/implementations/heavy/kuwahara.c
#include <stddef.h>
#include <stdint.h>
#include <stdlib.h>

typedef struct {
  float variance;
  float r;
  float g;
  float b;
} RegionStats_t;

static inline RegionStats_t get_stats(const uint8_t *pixels, int x1, int y1,
                                      int x2, int y2, int width, int height) {
  uint32_t r_sum = 0, g_sum = 0, b_sum = 0;
  uint32_t r_sq_sum = 0, g_sq_sum = 0, b_sq_sum = 0;
  uint32_t count = 0;

  for (int y = y1; y <= y2; y++) {
    int py = y;
    if (py < 0)
      py = 0;
    else if (py >= height)
      py = height - 1;
    size_t row_base = (size_t)py * width;

    for (int x = x1; x <= x2; x++) {
      int px = x;
      if (px < 0)
        px = 0;
      else if (px >= width)
        px = width - 1;

      size_t idx = (row_base + px) * PX_SIZE;

      uint32_t r = pixels[idx];
      uint32_t g = pixels[idx + 1];
      uint32_t b = pixels[idx + 2];

      r_sum += r;
      g_sum += g;
      b_sum += b;
      r_sq_sum += r * r;
      g_sq_sum += g * g;
      b_sq_sum += b * b;
      count++;
    }
  }

  float f_count = (float)count;
  float r_mean = (float)r_sum / f_count;
  float g_mean = (float)g_sum / f_count;
  float b_mean = (float)b_sum / f_count;

  float variance = ((float)r_sq_sum / f_count - r_mean * r_mean) +
                   ((float)g_sq_sum / f_count - g_mean * g_mean) +
                   ((float)b_sq_sum / f_count - b_mean * b_mean);

  RegionStats_t stats = {variance, r_mean, g_mean, b_mean};
  return stats;
}

void apply_kuwahara(uint8_t *pixels, uint8_t *output, size_t width,
                    size_t height, uint8_t radius) {
  if (radius < 1)
    return;

  int w = (int)width;
  int h = (int)height;
  int r = (int)radius;

  for (size_t y = 0; y < height; y++) {
    size_t row_base = y * width;
    for (size_t x = 0; x < width; x++) {

      int ix = (int)x;
      int iy = (int)y;

      RegionStats_t q1 = get_stats(pixels, ix - r, iy - r, ix, iy, w, h);
      RegionStats_t q2 = get_stats(pixels, ix, iy - r, ix + r, iy, w, h);
      RegionStats_t q3 = get_stats(pixels, ix - r, iy, ix, iy + r, w, h);
      RegionStats_t q4 = get_stats(pixels, ix, iy, ix + r, iy + r, w, h);

      float min_var = q1.variance;
      RegionStats_t best = q1;

      if (q2.variance < min_var) {
        min_var = q2.variance;
        best = q2;
      }
      if (q3.variance < min_var) {
        min_var = q3.variance;
        best = q3;
      }
      if (q4.variance < min_var) {
        best = q4;
      }

      size_t dest = (row_base + x) * PX_SIZE;

      output[dest] = clamp_to_u8_float(best.r);
      output[dest + 1] = clamp_to_u8_float(best.g);
      output[dest + 2] = clamp_to_u8_float(best.b);
      output[dest + 3] = pixels[dest + 3];
    }
  }
}

File: src/shared/lib/image-processing/heavy/kuwahara.ts
import type { FilterProcessFn } from "../types";

export const applyKuwahara: FilterProcessFn = (
  pixels,
  width,
  height,
  options
) => {
  const radius = options?.radius || 6;
  const output = new Uint8ClampedArray(pixels.length);

  const currentStats = new Float32Array(4);

  const evalRegion = (x1: number, y1: number, x2: number, y2: number) => {
    let rSum = 0,
      gSum = 0,
      bSum = 0;
    let rSqSum = 0,
      gSqSum = 0,
      bSqSum = 0;
    let count = 0;

    for (let y = y1; y <= y2; y++) {
      // inline clamp by Y
      let py = y;
      if (py < 0) py = 0;
      else if (py >= height) py = height - 1;
      const rowBase = py * width;

      for (let x = x1; x <= x2; x++) {
        // inline clamp by X
        let px = x;
        if (px < 0) px = 0;
        else if (px >= width) px = width - 1;

        const idx = (rowBase + px) << 2; // << 2 = * 4

        const r = pixels[idx];
        const g = pixels[idx + 1];
        const b = pixels[idx + 2];

        rSum += r;
        gSum += g;
        bSum += b;
        rSqSum += r * r;
        gSqSum += g * g;
        bSqSum += b * b;
        count++;
      }
    }

    const rMean = rSum / count;
    const gMean = gSum / count;
    const bMean = bSum / count;

    const variance =
      rSqSum / count -
      rMean * rMean +
      (gSqSum / count - gMean * gMean) +
      (bSqSum / count - bMean * bMean);

    currentStats[0] = variance;
    currentStats[1] = rMean;
    currentStats[2] = gMean;
    currentStats[3] = bMean;
  };

  for (let y = 0; y < height; y++) {
    const rowBase = y * width;
    for (let x = 0; x < width; x++) {
      // top-left
      evalRegion(x - radius, y - radius, x, y);
      let minVar = currentStats[0];
      let bestR = currentStats[1];
      let bestG = currentStats[2];
      let bestB = currentStats[3];

      // top-right
      evalRegion(x, y - radius, x + radius, y);
      if (currentStats[0] < minVar) {
        minVar = currentStats[0];
        bestR = currentStats[1];
        bestG = currentStats[2];
        bestB = currentStats[3];
      }

      // bottom-left
      evalRegion(x - radius, y, x, y + radius);
      if (currentStats[0] < minVar) {
        minVar = currentStats[0];
        bestR = currentStats[1];
        bestG = currentStats[2];
        bestB = currentStats[3];
      }

      // bottom-right
      evalRegion(x, y, x + radius, y + radius);
      if (currentStats[0] < minVar) {
        bestR = currentStats[1];
        bestG = currentStats[2];
        bestB = currentStats[3];
      }

      const dest = (rowBase + x) << 2;
      output[dest] = bestR;
      output[dest + 1] = bestG;
      output[dest + 2] = bestB;
      output[dest + 3] = pixels[dest + 3];
    }
  }

  return output;
};
